firmware/stellar.c

```diff
@@ -0,0 +1,876 @@
+}
+
+/*
+ * Returns a uint32 formatted as a string
+ */
+void stellar_format_uint32(uint64_t number, char *out, size_t outlen)
+{
+    bignum256 bn_number;
+    bn_read_uint32(number, &bn_number);
+    bn_format(&bn_number, NULL, NULL, 0, 0, false, out, outlen);
+}
+
+/*
+ * Returns a uint64 formatted as a string
+ */
+void stellar_format_uint64(uint64_t number, char *out, size_t outlen)
+{
+    bignum256 bn_number;
+    bn_read_uint64(number, &bn_number);
+    bn_format(&bn_number, NULL, NULL, 0, 0, false, out, outlen);
+}
+
+/*
+ * Breaks a 56 character address into 3 lines of lengths 16, 20, 20
+ * This is to allow a small label to be prepended to the first line
+ */
+const char **stellar_lineBreakAddress(uint8_t *addrbytes)
+{
+    char str_fulladdr[56+1];
+    static char rows[3][20+1];
+
+    memset(rows, 0, sizeof(rows));
+
+    // get full address string
+    stellar_publicAddressAsStr(addrbytes, str_fulladdr, sizeof(str_fulladdr));
+
+    // Break it into 3 lines
+    strlcpy(rows[0], str_fulladdr + 0, 17);
+    strlcpy(rows[1], str_fulladdr + 16, 21);
+    strlcpy(rows[2], str_fulladdr + 16 + 20, 21);
+
+    static const char *ret[3] = { rows[0], rows[1], rows[2] };
+    return ret;
+}
+
+size_t stellar_publicAddressAsStr(uint8_t *bytes, char *out, size_t outlen)
+{
+    // version + key bytes + checksum
+    uint8_t keylen = 1 + 32 + 2;
+    uint8_t bytes_full[keylen];
+    bytes_full[0] = 6 << 3; // 'G'
+
+    memcpy(bytes_full + 1, bytes, 32);
+
+    // Last two bytes are the checksum
+    uint16_t checksum = stellar_crc16(bytes_full, 33);
+    bytes_full[keylen-2] = checksum & 0x00ff;
+    bytes_full[keylen-1] = (checksum>>8) & 0x00ff;
+
+    base32_encode(bytes_full, keylen, out, outlen, BASE32_ALPHABET_RFC4648);
+
+    // Public key will always be 56 characters
+    return 56;
+}
+
+/*
+ * CRC16 implementation compatible with the Stellar version
+ * Ported from this implementation: http://introcs.cs.princeton.edu/java/61data/CRC16CCITT.java.html
+ * Initial value changed to 0x0000 to match Stellar
+ */
+uint16_t stellar_crc16(uint8_t *bytes, uint32_t length)
+{
+    // Calculate checksum for existing bytes
+    uint16_t crc = 0x0000;
+    uint16_t polynomial = 0x1021;
+    uint32_t i;
+    uint8_t bit;
+    uint8_t byte;
+    uint8_t bitidx;
+    uint8_t c15;
+
+    for (i=0; i < length; i++) {
+        byte = bytes[i];
+        for (bitidx=0; bitidx < 8; bitidx++) {
+            bit = ((byte >> (7 - bitidx) & 1) == 1);
+            c15 = ((crc >> 15 & 1) == 1);
+            crc <<= 1;
+            if (c15 ^ bit) crc ^= polynomial;
+        }
+    }
+
+    return crc & 0xffff;
+}
+
+/*
+ * Writes 32-byte public key to out
+ */
+void stellar_getPubkeyAtIndex(uint32_t index, uint8_t *out, size_t outlen)
+{
+    if (outlen < 32) return;
+
+    HDNode *node = stellar_deriveNode(index);
+
+    ed25519_public_key stellarPubkeyBytes;
+    ed25519_publickey(node->private_key, stellarPubkeyBytes);
```

Comment:
You have already set `node->public_key` in `stellar_deriveNode`, unless I'm mistaken.